File: timerx-example/src/main/java/com/arsvechkarev/timerxexample/TimerFragment.java
package com.arsvechkarev.timerxexample;


import static java.util.concurrent.TimeUnit.SECONDS;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import timerx.Timer;
import timerx.TimerBuilder;

/**
 * Example of using timer
 */
public class TimerFragment extends Fragment {

  private Timer timer;

  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    return inflater.inflate(R.layout.fragment_timer, container, false);
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    Button buttonStart = view.findViewById(R.id.btn_start);
    Button buttonStop = view.findViewById(R.id.btn_stop);
    Button buttonReset = view.findViewById(R.id.btn_reset);

    timer = new TimerBuilder()
        .startFormat("SS:LL")
        .startTime(25, SECONDS)
        .actionWhen(18, SECONDS, () -> showToast("18s left"))
        .actionWhen(10, SECONDS, () -> showToast("10s left"))
        .actionWhen(5, SECONDS, () -> showToast("5s left"))
        .changeFormatWhen(10, SECONDS, "10: SS:LLL")
        .changeFormatWhen(5, SECONDS, "5: SS:LLLLL")
        .onTick(time -> {
          Log.d("qwerty", "time = " + time);
        })
        .build();

    buttonStart.setOnClickListener(v -> {
      timer.start();
    });

    buttonStop.setOnClickListener(v -> {
      timer.stop();
      showToast("Remaining time in seconds = " + timer.getRemainingTimeIn(SECONDS));
    });

    buttonReset.setOnClickListener(v -> {
      timer.reset();
    });

  }

  private void showToast(String text) {
    Toast.makeText(getContext(), text, Toast.LENGTH_LONG).show();
  }
}

File: timerx-example/src/main/java/com/arsvechkarev/timerxexample/TimerFragment.kt
package com.arsvechkarev.timerxexample

import android.os.Bundle
import android.os.CountDownTimer
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.fragment_timer.btn_reset
import kotlinx.android.synthetic.main.fragment_timer.btn_set_time
import kotlinx.android.synthetic.main.fragment_timer.btn_start
import kotlinx.android.synthetic.main.fragment_timer.btn_stop
import kotlinx.android.synthetic.main.fragment_timer.text_time
import timerx.Timer
import timerx.buildTimer
import java.util.concurrent.TimeUnit

/**
 * Example of using a timer
 */
class TimerFragment : Fragment() {
  
  private var cTimer: CountDownTimer? = null
  private lateinit var timer: Timer
  
  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    return inflater.inflate(R.layout.fragment_timer, container, false)
  }
  
  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    
    timer = buildTimer {
      startFormat("30s -> MM:SS")
      startTime(15, TimeUnit.SECONDS)
      useExactDelay(true)
      changeFormatWhen(10, TimeUnit.SECONDS, "10s -> SS")
//      changeFormatWhen(5, TimeUnit.SECONDS, "5#S -> SS:L")
//      actionWhen(20, TimeUnit.SECONDS) { showToast("20 seconds left") }
//      actionWhen(25, TimeUnit.SECONDS) { showToast("25 seconds left") }
      onTick { millis, formattedTime ->
        text_time.text = formattedTime
        Log.i("Timer", "Remaining time = $millis")
      }
      onFinish { showToast("Finished!") }
    }
    
    text_time.text = timer.formattedStartTime
    
    btn_start.setOnClickListener { timer.start() }
    
    btn_stop.setOnClickListener {
      timer.stop()
    }
    
    btn_set_time.setOnClickListener {
      timer.setTime(15, TimeUnit.SECONDS)
      // Since call to setTimeTo() does not result in invoking onTick() callback,
      // we are setting time to textView manually so that it appears there even if
      // timer is paused now
      text_time.text = timer.remainingFormattedTime
    }
    
    btn_reset.setOnClickListener {
      text_time.text = timer.formattedStartTime
      timer.reset()
    }
  }
  
  override fun onDestroyView() {
    super.onDestroyView()
    timer.release()
  }
}
